cudax/test/containers/async_buffer/test_resources.h

```diff
@@ -0,0 +1,216 @@
+  Resource resource_{};
+  std::unordered_multimap<std::size_t, void*> available_allocations_{};
+  std::unordered_multimap<std::size_t, void*> used_allocations_{};
+
+public:
+  caching_resource() = default;
+  caching_resource(const caching_resource& other)
+      : resource_(other.resource_)
+      , available_allocations_()
+      , used_allocations_()
+  {}
+  caching_resource& operator=(const caching_resource& other)
+  {
+    CUDAX_CHECK(used_allocations_.empty());
+    resource_ = other.resource_;
+  }
+  caching_resource(caching_resource&& other)
+      : resource_(cuda::std::move(other.resource_))
+      , available_allocations_(cuda::std::move(other.available_allocations_))
+      , used_allocations_(cuda::std::move(other.used_allocations_))
+  {}
+  caching_resource& operator=(caching_resource&& other)
+  {
+    CUDAX_CHECK(used_allocations_.empty());
+    resource_              = cuda::std::move(other.resource_);
+    available_allocations_ = cuda::std::move(other.available_allocations_);
+    used_allocations_      = cuda::std::move(other.used_allocations_);
+  }
+
+  ~caching_resource()
+  {
+    CUDAX_CHECK(used_allocations_.empty());
+    for (const auto& elem : available_allocations_)
+    {
+      resource_.deallocate(elem.second, elem.first, alignof(int));
+    }
+  }
+
+  void* allocate(std::size_t size, std::size_t alignment)
+  {
+    if (available_allocations_.count(size) == 0)
+    {
+      return used_allocations_.emplace(size, resource_.allocate(size, alignment))->second;
+    }
+    else
+    {
+      return used_allocations_.insert(available_allocations_.extract(size))->second;
+    }
+  }
+
+  void* allocate_async(std::size_t size, std::size_t alignment, ::cuda::stream_ref stream)
+  {
+    if (available_allocations_.count(size) == 0)
+    {
+      return used_allocations_.emplace(size, resource_.allocate_async(size, alignment, stream))->second;
+    }
+    else
+    {
+      return used_allocations_.insert(available_allocations_.extract(size))->second;
+    }
+  }
+
+  void deallocate(void* ptr, std::size_t size, std::size_t)
+  {
+    CUDAX_CHECK(used_allocations_.count(size) != 0);
+    auto range = used_allocations_.equal_range(size);
+    for (auto curr = range.first; curr != range.second; ++curr)
+    {
+      if (curr->second == ptr)
+      {
+        available_allocations_.insert(used_allocations_.extract(curr));
+        return;
+      }
+    }
+    CUDAX_CHECK(false);
+  }
+
+  void deallocate_async(void* ptr, std::size_t size, std::size_t alignment, ::cuda::stream_ref)
+  {
+    deallocate(ptr, size, alignment);
+  }
+
+  bool operator==(const caching_resource& other) const
+  {
+    return resource_ == other.resource_;
+  }
+  bool operator!=(const caching_resource& other) const
+  {
+    return resource_ != other.resource_;
+  }
+
+  _CCCL_TEMPLATE(class Property)
+  _CCCL_REQUIRES(cuda::has_property<Resource, Property>)
+  friend void get_property(const caching_resource&, Property) noexcept {}
```

Comment:
i have distant and foggy memories that nvcc or nvc++ has problems with constrained friends, and that (maybe?) using the trailing requires form fixes it. but the tests are green so maybe i'm misremembering.